causallearn/utils/KCI/KCI.py

```diff
@@ -453,9 +465,21 @@ def KCI_V_statistic(self, Kx, Ky, Kzx, Kzy):
         Vstat: KCI v statistics
         KxR: centralized kernel matrix for data_x (nxn)
         KyR: centralized kernel matrix for data_y (nxn)
+
+        [Updated @Haoyue 06/24/2022]
+        1. Kx, Ky, Kzx, Kzy are all symmetric matrices.
+            - Kx, Ky are with diagonal elements of 1 (because of exp(-0.5 * sq_dists * self.width)).
+            - If (self.kernelZ == 'Gaussian' and self.use_gp), then Kzx (Kzy) has all the same diagonal elements (not necessarily 1).
+              Otherwise Kzx, Kzy are with diagonal elements of 1.
+        2. If not (self.kernelZ == 'Gaussian' and self.use_gp): assert (Kzx == Kzy).all()
+           With this we could save one repeated calculation of pinv(Kzy+\epsilonI), which consumes most time.
         """
-        KxR = Kernel.center_kernel_matrix_regression(Kx, Kzx, self.epsilon_x)
-        KyR = Kernel.center_kernel_matrix_regression(Ky, Kzy, self.epsilon_y)
+        KxR, Rzx = Kernel.center_kernel_matrix_regression(Kx, Kzx, self.epsilon_x)
+        if self.epsilon_x != self.epsilon_y or (self.kernelZ == 'Gaussian' and self.use_gp):
+            KyR, _ = Kernel.center_kernel_matrix_regression(Ky, Kzy, self.epsilon_y)
+        else:
+            # assert np.all(Kzx == Kzy), 'Kzx and Kzy are the same'
```

Comment:
why this assertion is commented? We should have assertions whenever possible

Reply:
Just for speed. `np.all(Kzx == Kzy)` takes O(n^2) time to traverse each elements (just the same as the key function part e.g. `center_kernel_matrix` after optimization).

I had assertions when writing and testing. But I tried to remove them for deployment. What do you think?

Reply:
cool, sounds good! Will any user input break this assertion? If so, maybe we can still add this assertion to make sure user gives the correct input.

Reply:
To the user end, any data input or parameters choice will not break this assertion, i think.